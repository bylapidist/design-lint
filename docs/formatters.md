# Formatters

`@lapidist/design-lint` uses formatter functions to turn lint results into human-
or machine-readable output. A formatter receives an array of `LintResult`
objects and an optional `useColor` flag and returns a string to print.

## Built-in formatters

- `stylish` – default, colorized summary intended for terminals.
- `json` – raw JSON results, useful for piping to other tools.
- `sarif` – emits a [SARIF 2.1.0](https://sarifweb.azurewebsites.net/) log for CI systems.

Select a formatter with the `--format` CLI option or the `getFormatter(name)` API.

## Creating a custom formatter

You can supply your own formatter module without modifying
`@lapidist/design-lint`.

1. **Create the formatter module**:

   ```ts
   // minimal-formatter.ts
   import type { LintResult } from '@lapidist/design-lint';

   export default function minimal(results: LintResult[]): string {
     return results.map((r) => `${r.filePath}: ${r.messages.length}`).join('\n');
   }
   ```

2. **Use the formatter from the CLI**:

   ```bash
   npx design-lint src --format ./minimal-formatter.ts
   ```

3. **Or from the API**:

   ```ts
   import { getFormatter } from '@lapidist/design-lint';

   const formatter = await getFormatter('./minimal-formatter.ts');
   console.log(formatter(results));
   ```

Formatters receive a `useColor` boolean as the second argument. Respect this
flag if your formatter supports colored output. For inspiration, see the
built-in formatter implementations in [`src/formatters/`](../src/formatters/).
