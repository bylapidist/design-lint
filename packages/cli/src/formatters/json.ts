import type { LintResult } from '@lapidist/design-lint-core';

export function jsonFormatter(results: LintResult[], _useColor = true): string {
  void _useColor;
  return JSON.stringify(results, null, 2);
}
