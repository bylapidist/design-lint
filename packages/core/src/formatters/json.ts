import type { LintResult } from '../core/types.js';

export function jsonFormatter(results: LintResult[], _useColor = true): string {
  void _useColor;
  return JSON.stringify(results, null, 2);
}
