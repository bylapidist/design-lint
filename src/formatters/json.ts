import type { LintResult } from '../core/types';

export function jsonFormatter(results: LintResult[]): string {
  return JSON.stringify(results, null, 2);
}
