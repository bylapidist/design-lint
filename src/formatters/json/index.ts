/**
 * @packageDocumentation
 *
 * JSON formatter for machine-readable lint results.
 *
 * Serializes {@link LintResult} arrays into a human-friendly JSON string.
 */
export { jsonFormatter } from './json-formatter.js';
