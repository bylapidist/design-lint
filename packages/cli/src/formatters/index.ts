import path from 'node:path';
import { pathToFileURL } from 'node:url';
import type { LintResult } from '@lapidist/design-lint-core';
import { stylish } from './stylish.js';
import { jsonFormatter } from './json.js';
import { sarifFormatter } from './sarif.js';

type Formatter = (results: LintResult[], useColor?: boolean) => string;

/**
 * Retrieve a formatter by name.
 * @param name Formatter identifier or module path.
 * @returns Promise resolving to a formatter function.
 */
export async function getFormatter(name: string): Promise<Formatter> {
  switch (name) {
    case 'stylish':
      return stylish;
    case 'json':
      return jsonFormatter;
    case 'sarif':
      return sarifFormatter;
    default: {
      try {
        const resolved =
          path.isAbsolute(name) ||
          name.startsWith('./') ||
          name.startsWith('../')
            ? pathToFileURL(path.resolve(process.cwd(), name)).href
            : import.meta.resolve(
                name,
                pathToFileURL(path.join(process.cwd(), 'index.js')).href,
              );
        const mod = (await import(resolved)) as unknown;
        const formatter = resolveFormatter(mod);
        if (!formatter) {
          throw new Error();
        }
        return formatter;
      } catch {
        throw new Error(`Unknown formatter: ${name}`);
      }
    }
  }
}

function resolveFormatter(mod: unknown): Formatter | undefined {
  if (isFormatter(mod)) {
    return mod;
  }
  if (isRecord(mod)) {
    if (isFormatter(mod.default)) {
      return mod.default;
    }
    if (isFormatter(mod.formatter)) {
      return mod.formatter;
    }
  }
  return undefined;
}

function isFormatter(value: unknown): value is Formatter {
  return typeof value === 'function';
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}
