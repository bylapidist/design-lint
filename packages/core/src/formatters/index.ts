import { pathToFileURL } from 'node:url';
import type { LintResult } from '../core/types.js';
import { stylish } from './stylish.js';
import { jsonFormatter } from './json.js';
import { sarifFormatter } from './sarif.js';
import type { Env } from '@lapidist/design-lint-shared';
import { nodeEnv } from '@lapidist/design-lint-shared';

type Formatter = (results: LintResult[], useColor?: boolean) => string;

/**
 * Retrieve a formatter by name.
 * @param name Formatter identifier or module path.
 * @param env Environment adapters.
 * @returns Promise resolving to a formatter function.
 */
export async function getFormatter(
  name: string,
  env: Env = nodeEnv,
): Promise<Formatter> {
  switch (name) {
    case 'stylish':
      return stylish;
    case 'json':
      return jsonFormatter;
    case 'sarif':
      return sarifFormatter;
    default: {
      try {
        const resolved =
          env.path.isAbsolute(name) ||
          name.startsWith('./') ||
          name.startsWith('../')
            ? pathToFileURL(env.path.resolve(process.cwd(), name)).href
            : import.meta.resolve(
                name,
                pathToFileURL(env.path.join(process.cwd(), 'index.js')).href,
              );
        const mod = (await import(resolved)) as unknown;
        const formatter = resolveFormatter(mod);
        if (!formatter) {
          throw new Error();
        }
        return formatter;
      } catch {
        throw new Error(`Unknown formatter: ${name}`);
      }
    }
  }
}

function resolveFormatter(mod: unknown): Formatter | undefined {
  if (isFormatter(mod)) {
    return mod;
  }
  if (isRecord(mod)) {
    if (isFormatter(mod.default)) {
      return mod.default;
    }
    if (isFormatter(mod.formatter)) {
      return mod.formatter;
    }
  }
  return undefined;
}

function isFormatter(value: unknown): value is Formatter {
  return typeof value === 'function';
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}
