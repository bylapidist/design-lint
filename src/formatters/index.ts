import type { LintResult } from '../core/types';
import { stylish } from './stylish';
import { jsonFormatter } from './json';
import { sarifFormatter } from './sarif';

type Formatter = (results: LintResult[]) => string;

export function getFormatter(name: string): Formatter {
  switch (name) {
    case 'json':
      return jsonFormatter;
    case 'sarif':
      return sarifFormatter;
    default:
      return stylish;
  }
}
