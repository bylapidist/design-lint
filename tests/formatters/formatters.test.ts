import { stylish } from '../../src/formatters/stylish';
import { jsonFormatter } from '../../src/formatters/json';
import { sarifFormatter } from '../../src/formatters/sarif';
import type { LintResult } from '../../src/core/types';

const sample: LintResult[] = [
  {
    filePath: 'file.ts',
    messages: [
      { ruleId: 'rule', message: 'msg', severity: 'error', line: 1, column: 1 },
    ],
  },
];

test('stylish', () => {
  expect(stylish(sample, false)).toMatchSnapshot();
});

test('json', () => {
  expect(jsonFormatter(sample)).toMatchSnapshot();
});

test('sarif', () => {
  expect(sarifFormatter(sample)).toMatchSnapshot();
});
