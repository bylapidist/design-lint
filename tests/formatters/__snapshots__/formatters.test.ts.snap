// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json 1`] = `
"[
  {
    "filePath": "file.ts",
    "messages": [
      {
        "ruleId": "rule",
        "message": "msg",
        "severity": "error",
        "line": 1,
        "column": 1
      }
    ]
  }
]"
`;

exports[`sarif 1`] = `
"{
  "version": "2.1.0",
  "runs": [
    {
      "tool": {
        "driver": {
          "name": "design-lint",
          "informationUri": "https://github.com/"
        }
      },
      "results": [
        {
          "ruleId": "rule",
          "message": {
            "text": "msg"
          },
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "file.ts"
                },
                "region": {
                  "startLine": 1,
                  "startColumn": 1
                }
              }
            }
          ]
        }
      ]
    }
  ]
}"
`;

exports[`stylish 1`] = `
"file.ts
  1:1  error  msg  rule
"
`;
